### DISCUSSION TITLE
Use Prism as the default parser?

### DISCUSSION THREAD

**Discussion by exterm** [Shopify/packwerk#387] [Ideas] [answered]
Friends,

last time I profiled Packwerk a few years ago, about 2/3 of the execution time was spent parsing. A coworker just recently profiled it again and basically this still seems to be the case.

I [tried to write a new parser based on `ripper`](https://github.com/Shopify/packwerk/pull/148) in 2021, when I was still at Shopify. However, that turned out to be a lot of work. Back then, I was hoping for @kddnewton 's work on a new parser to bear fruit, so that we can use it for packwerk at some point. 

Well, [prism is now a thing](https://github.com/ruby/prism). Would it make sense to swap out the currently used whitequark parser for prism?

@kddnewton how much faster do you expect Prism to be compared with the whitequark parser? I couldn't find any benchmarks.

If this makes sense, I might take a stab at it - provided I can get someone to pay me for the time spent on it

**Comment by kddnewton** [comment, marked as answer]
I would expect it to be significantly faster. As an example, I used it as the basis of the parser gem for running rubocop, and the benchmarks are here: https://github.com/kddnewton/parser-prism?tab=readme-ov-file#benchmarks.

I would be very supportive of this work!

**Reply by exterm** [reply]
Is `parser-prism` production ready or are you pllaning to get it there?

If we could just swap that in without having to rewrite the code currently specific to `whitequark/parser`, that sounds like a good option

**Reply by exterm** [reply]
I tried `parser-prism` and I only get a single failing test https://github.com/Shopify/packwerk/pull/388

I'll look into that one

**Reply by exterm** [reply]
The problem seems to be that `Parser::Prism` raises a `StandardError` [here](https://github.com/kddnewton/parser-prism/blob/2261e791846a6df662f4a51197b4087619cdce35/lib/parser/prism/compiler.rb#L1051), whereas `Parser::CurrentRuby` raises a `Parser::SyntaxError` in that case.

`Parser::Prism` raises a more specific error on `parser-prism`'s main branch. We could cath that one explicitly. However, it seems for compatibility you may want to raise the same exceptions as `Parser::CurrentRuby`?

**Reply by exterm** [reply]
Let's continue the discussion on the PR

**Comment by gmcgibbon** [comment]
I believe we've discussed this internally at some point. I'm also very much in favour of this change, but we would have to update [`NodeHelpers`](https://github.com/Shopify/packwerk/blob/main/lib/packwerk/node_helpers.rb) among other things that are keyed to the parser api.

**Reply by exterm** [reply]
It looks like we might not have to if we use `parser-prism`? See [comment above](https://github.com/Shopify/packwerk/discussions/387#discussioncomment-8339555).
